elfpy/utils/apeworx_integrations.py

```diff
@@ -0,0 +1,250 @@
+"""Helper functions for integrating the sim repo with solidity contracts via Apeworx"""
+
+from __future__ import annotations
+from typing import TYPE_CHECKING, Any
+from enum import IntEnum
+
+import logging
+
+# TODO: apeworx is not worxing with github actions when it is listed in requirements
+# and pyright doesn't like imports that aren't also in requirements.
+# pylint: disable=import-error
+import ape  # type: ignore[reportMissingImports]
+
+if TYPE_CHECKING:
+    from ape.types import ContractLog  # type: ignore[reportMissingImports]
+    from ape.api.accounts import AccountAPI  # type: ignore[reportMissingImports]
+    from ape_ethereum.transactions import Receipt  # type: ignore[reportMissingImports]
+    from ape.contracts.base import ContractInstance  # type: ignore[reportMissingImports]
+
+
+class AssetIdPrefix(IntEnum):
+    r"""The asset ID is used to encode the trade type in a transaction receipt"""
+    LONG = 0
+    SHORT = 1
+    WITHDRAWAL_SHARE = 2
+
+
+def encode_asset_id(prefix: int, timestamp: int) -> int:
+    r"""Encodes a prefix and a timestamp into an asset ID.
+
+    Asset IDs are used so that LP, long, and short tokens can all be represented
+    in a single MultiToken instance. The zero asset ID indicates the LP token.
+
+    Encode the asset ID by left-shifting the prefix by 248 bits,
+    then bitwise-or-ing the result with the timestamp.
+
+    Argments
+    --------
+    prefix: int
+        A one byte prefix that specifies the asset type.
+    timestamp: int
+        A timestamp associated with the asset.
+
+    Returns
+    -------
+    int
+        The asset ID.
+    """
+    timestamp_mask = (1 << 248) - 1
+    if timestamp > timestamp_mask:
+        raise ValueError("Invalid timestamp")
+    return (prefix << 248) | timestamp
+
+
+def decode_asset_id(asset_id: int) -> tuple[int, int]:
+    r"""Decodes a transaction asset ID into its constituent parts of an identifier, data, and a timestamp.
+
+    First calculate the prefix mask by left-shifting 1 by 248 bits and subtracting 1 from the result.
+    This gives us a bit-mask with 248 bits set to 1 and the rest set to 0.
+    Then apply this mask to the input ID using the bitwise-and operator `&` to extract
+    the lower 248 bits as the timestamp.
+
+    Arguments
+    ---------
+    asset_id: int
+        Encoded ID from a transaction. It is a concatenation, [identifier: 8 bits][timestamp: 248 bits]
+
+    Returns
+    -------
+    tuple[int, int]
+        identifier, timestamp
+    """
+    prefix_mask = (1 << 248) - 1
+    prefix = asset_id >> 248  # shr 248 bits
+    timestamp = asset_id & prefix_mask  # apply the prefix mask
+    return prefix, timestamp
+
+
+def get_transaction_trade_event(tx_receipt: Receipt) -> ContractLog:
```

Comment:
we should be absolutely clear that this is just for the hyperdrive contract.  could incidentally work for others, but not for sure.